File: server/apps/nexotype/schemas/knowledge_graph/therapeutic_efficacy_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class TherapeuticEfficacyCreate(BaseModel):
    """Schema for creating a new therapeutic efficacy"""
    asset_id: int
    indication_id: int | None = None
    phenotype_id: int | None = None
    biomarker_id: int | None = None
    direction: str
    magnitude: str | None = None


class TherapeuticEfficacyUpdate(BaseModel):
    """Schema for updating a therapeutic efficacy"""
    asset_id: int = None
    indication_id: int | None = None
    phenotype_id: int | None = None
    biomarker_id: int | None = None
    direction: str = None
    magnitude: str | None = None


# ==========================================
# Response Schemas
# ==========================================

class TherapeuticEfficacyDetail(BaseModel):
    """Schema for therapeutic efficacy details"""
    id: int
    asset_id: int
    indication_id: int | None = None
    phenotype_id: int | None = None
    biomarker_id: int | None = None
    direction: str
    magnitude: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class TherapeuticEfficacyListResponse(BaseModel):
    """Response schema for listing therapeutic efficacies"""
    success: bool
    data: list[TherapeuticEfficacyDetail] | None = None
    count: int = 0
    error: str | None = None


class TherapeuticEfficacyResponse(BaseModel):
    """Response schema for single therapeutic efficacy operations"""
    success: bool
    data: TherapeuticEfficacyDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/knowledge_graph/therapeutic-efficacy.schemas.ts
import { z } from 'zod';

// ==========================================
// Enums
// ==========================================

export const DirectionEnum = z.enum(['Increases', 'Decreases', 'Ameliorates']);

// ==========================================
// Options (for Select dropdowns)
// ==========================================

/** TherapeuticEfficacy.direction */
export const DIRECTION_OPTIONS = [
  { label: 'Increases', value: 'Increases' },
  { label: 'Decreases', value: 'Decreases' },
  { label: 'Ameliorates', value: 'Ameliorates' },
] as const;

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const TherapeuticEfficacySchema = z.object({
  id: z.number(),
  asset_id: z.number().int(),
  indication_id: z.number().int().nullable().optional(),
  phenotype_id: z.number().int().nullable().optional(),
  biomarker_id: z.number().int().nullable().optional(),
  direction: DirectionEnum,
  magnitude: z.string().max(50).nullable().optional(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateTherapeuticEfficacySchema = z.object({
  asset_id: z.number().int(),
  indication_id: z.number().int().nullable().optional(),
  phenotype_id: z.number().int().nullable().optional(),
  biomarker_id: z.number().int().nullable().optional(),
  direction: DirectionEnum,
  magnitude: z.string().max(50).nullable().optional(),
});

export const UpdateTherapeuticEfficacySchema = z.object({
  asset_id: z.number().int().optional(),
  indication_id: z.number().int().nullable().optional(),
  phenotype_id: z.number().int().nullable().optional(),
  biomarker_id: z.number().int().nullable().optional(),
  direction: DirectionEnum.optional(),
  magnitude: z.string().max(50).nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type TherapeuticEfficacy = z.infer<typeof TherapeuticEfficacySchema>;
export type CreateTherapeuticEfficacy = z.infer<typeof CreateTherapeuticEfficacySchema>;
export type UpdateTherapeuticEfficacy = z.infer<typeof UpdateTherapeuticEfficacySchema>;
export type Direction = z.infer<typeof DirectionEnum>;

// ==========================================
// Response Types
// ==========================================

export type TherapeuticEfficacyResponse = {
  success: boolean;
  data?: TherapeuticEfficacy;
  error?: string;
};

export type TherapeuticEfficaciesResponse = {
  success: boolean;
  data?: TherapeuticEfficacy[];
  count?: number;
  error?: string;
};
